Source/JavaScriptCore/runtime/JSStringJoiner.h

```diff
@@ -217,4 +217,49 @@ ALWAYS_INLINE void JSStringJoiner::appendNumber(VM& vm, double value)
         append8Bit(vm.numericStrings.add(value));
 }
 
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+
+// Avoids the overhead of accumulating intermediate vectors of values when
+// we're only joining existing strings.
+class JSOnlyStringsJoiner {
+public:
+    JSOnlyStringsJoiner(StringView separator)
+        : m_separator(separator)
+        , m_isAll8Bit(m_separator.is8Bit())
+    {
+    }
+
+    JSValue tryJoin(JSGlobalObject* globalObject, const WriteBarrier<Unknown>* data, unsigned length)
+    {
+        if (length == 1) {
+            JSValue value = data[0].get();
+            if (!value || !value.isString())
+                return JSValue();
+            return value;
+        }
+
+        for (size_t i = 0; i < length; ++i) {
+            JSValue value = data[i].get();
+            if (!value || !value.isString())
+                return JSValue();
+
+            JSString* string = asString(value);
+
+            m_accumulatedStringsLength += string->length();
+            m_isAll8Bit &= string->is8Bit();
+        }
+
+        return joinSlow(globalObject, data, length);
```

Comment:
Nit: Seems more like a `joinImpl` rather than `joinSlow` since you'll still call this in most cases.